magento/magento-cloud-docker
# MAGECLOUD-4894  [MTS] Add support of split DB on docker

https://magento2.atlassian.net/browse/MAGECLOUD-4894
### Description
Adds the support of Split Db

### Fixed Issues (if relevant)
1. https://magento2.atlassian.net/browse/MAGECLOUD-4894

### Manual testing scenarios
1. Clone any `[magento-cloud](https://github.com/magento/magento-cloud)` template
2. Run in the template directory  the cli command `composer update`
3. Run command `php vendor/bin/ece-docker build:compose`
   Expected results:
   The file `docker-compose.yaml` was generated without split Db services:
```shell
$ grep -E "(quote|sales)" ./docker-compose.yml

$ grep -E "(quote|sales)" ./.docker/config.php.dist

```
3. Add the next database services to the config file `.magento/services.yaml`:
```yaml
 ...
mysql-quote:
    type: mysql:10.2
    disk: 512
mysql-sales:
    type: mysql:10.2
    disk: 512
 ...
```
4. Add the next relationships to the config file `.magento.app.yaml`:
```yaml
relationships:
    ...
    database-quote: "mysql-quote:mysql"
    database-sales: "mysql-sales:mysql"
   ...
```
5. Run command `php vendor/bin/ece-docker build:compose`
   Expected results:
   The file `docker-compose.yaml` `.docker/config.php.dist` were generated with split Db services
6. Check `./docker-compose.yml`
```
$ cat  ./docker-compose.yml

...
services:
  ...
  db-quote:
    hostname: db-quote.magento2.docker
    image: 'mariadb:10.2'
    environment:
      - MYSQL_ROOT_PASSWORD=magento2
      - MYSQL_DATABASE=magento2
      - MYSQL_USER=magento2
      - MYSQL_PASSWORD=magento2
    ports:
      - '3306'
    volumes:
      - 'magento-db-quote:/var/lib/mysql'
      - 'docker-entrypoint-quote:/docker-entrypoint-initdb.d'
      - 'mariadb-conf:/etc/mysql/mariadb.conf.d'
      - 'docker-mnt:/mnt:delegated'
    networks:
      magento:
        aliases:
          - db-quote.magento2.docker
  db-sales:
    hostname: db-sales.magento2.docker
    image: 'mariadb:10.2'
    environment:
      - MYSQL_ROOT_PASSWORD=magento2
      - MYSQL_DATABASE=magento2
      - MYSQL_USER=magento2
      - MYSQL_PASSWORD=magento2
    ports:
      - '3306'
    volumes:
      - 'magento-db-sales:/var/lib/mysql'
      - 'docker-entrypoint-sales:/docker-entrypoint-initdb.d'
      - 'mariadb-conf:/etc/mysql/mariadb.conf.d'
      - 'docker-mnt:/mnt:delegated'
    networks:
      magento:
        aliases:
          - db-sales.magento2.docker
...
```
     
6. The file `./.docker/config.php.dist` was generated with next configurations:

```php
<?php

return [
    'MAGENTO_CLOUD_RELATIONSHIPS' => base64_encode(json_encode([
       ...
        'database-sales' => [
            [
                'host' => 'db-sales',
                'path' => 'magento2',
                'password' => 'magento2',
                'username' => 'magento2',
                'port' => '3306'
            ]
        ],
        'database-quote' => [
            [
                'host' => 'db-quote',
                'path' => 'magento2',
                'password' => 'magento2',
                'username' => 'magento2',
                'port' => '3306'
            ]
        ],
    ])),
   ...
];
```

7. Run `docker-compose up -d`
    Expected results:
    1. All services started successfully
8. Run command `docker-compose run deploy bash`
9. Run command in the container: ` echo $MAGENTO_CLOUD_RELATIONSHIPS | base64 -d | json_pp `
    Expected result:
```
{
...
   "database-quote" : [
      {
         "host" : "db-quote",
         "port" : "3306",
         "path" : "magento2",
         "username" : "magento2",
         "password" : "magento2"
      }
   ],
   "database-sales" : [
      {
         "path" : "magento2",
         "port" : "3306",
         "username" : "magento2",
         "host" : "db-sales",
         "password" : "magento2"
      }
   ],
...
}
 ```
10. Use credentials from the previous step to connect to DB services
```
mysql -h db-sales -u magento2 -pmagento2 magento2
```
```
mysql -h db-quote -u magento2 -pmagento2 magento2
```
11. Exit from container and run `docker-compose ps`
```
docker-compose ps
            Name                           Command                  State                Ports
-------------------------------------------------------------------------------------------------------
...
magento-cloud_db-quote_1        docker-entrypoint.sh mysqld      Up             0.0.0.0:32873->3306/tcp
magento-cloud_db-sales_1        docker-entrypoint.sh mysqld      Up             0.0.0.0:32874->3306/tcp
magento-cloud_db_1              docker-entrypoint.sh mysqld      Up             0.0.0.0:32872->3306/tcp
...
```
12. Try to connect to DB services from a host using port forwarding

### Contribution checklist
